<small><small><i>
All the IPython Notebooks in this lecture series by Dr. Milan Parmar are available @ **[GitHub](https://github.com/milaan9/04_Python_Functions/tree/main/002_Python_Functions_Built_in)**
</i></small></small>


# Python `round()`

The **`round()`** function returns a floating-point number rounded to the specified number of decimals.

**Syntax**:

```python
round(number, ndigits)
```


## `round()` Parameters

The **`round()`** function takes two parameters:

- **number** - the number to be rounded
- **ndigits (optional)** - number up to which the given number is rounded; defaults to 0


## Return Value from `round()`

- If **`ndigits`** is not provided, **`round()`** returns the nearest integer to the given number.
- If **`ndigits`** is given, **`round()`** returns the number rounded off to the **`ndigits`** digits.


In [1]:
# Example 1: How round() works in Python?

# for integers
print(round(10))

# for floating point
print(round(10.7))

# even choice
print(round(5.5))

10
11
6


In [2]:
# Example 2: Round a number to the given number of decimal places

print(round(2.665, 2))
print(round(2.675, 2))

2.67
2.67


> **Note:** The behavior of **`round()`** for floats can be surprising. Notice **`round(2.675, 2) `** gives **`2.67`** instead of the expected **`2.68`**. This is not a bug: it's a result of the fact that most decimal fractions can't be represented exactly as a float.
> When the decimal **`2.675`** is converted to a binary floating-point number, it's again replaced with a binary approximation, whose exact value is:

```python
2.67499999999999982236431605997495353221893310546875
```

Due to this, it is rounded down to 2.67.


If you're in a situation where this precision is needed, consider using the **`decimal`** module, which is designed for floating-point arithmetic:


In [3]:
# Example 3:

from decimal import Decimal

# normal float
num = 2.675
print(round(num, 2))

# using decimal.Decimal (passed float as string for precision)
num = Decimal('2.675')
print(round(num, 2))

2.67
2.68


The `round()` function in Python is a built-in function used to **round numbers** to a specified number of decimal places. It is commonly used for formatting floating-point numbers, reducing precision, or rounding to the nearest integer. Below is a comprehensive guide to the `round()` function, covering all its syntaxes and usage from basic to advanced.

---

## **Table of Contents**

1. **Basics of `round()`**
   - Rounding to the Nearest Integer
   - Rounding to a Specified Number of Decimal Places
2. **Advanced Usage of `round()`**
   - Rounding Negative Numbers
   - Rounding Ties (Halfway Cases)
   - Rounding with `Decimal` for Precision
3. **Practical Use Cases**
   - Financial Calculations
   - Formatting Output
   - Reducing Floating-Point Precision
4. **Advanced Topics**
   - Custom Rounding Functions
   - Rounding with `math` Module
   - Performance Considerations

---

## **1. Basics of `round()`**

### **Rounding to the Nearest Integer**

The simplest use of `round()` is to round a number to the nearest integer.

```python
# Example 1: Rounding positive numbers
print(round(3.14))  # Output: 3
print(round(3.75))  # Output: 4

# Example 2: Rounding negative numbers
print(round(-3.14))  # Output: -3
print(round(-3.75))  # Output: -4
```

### **Rounding to a Specified Number of Decimal Places**

You can specify the number of decimal places to round to using the second argument of `round()`.

```python
# Example 1: Rounding to 1 decimal place
print(round(3.14159, 1))  # Output: 3.1

# Example 2: Rounding to 2 decimal places
print(round(3.14159, 2))  # Output: 3.14

# Example 3: Rounding to 0 decimal places (same as rounding to integer)
print(round(3.14159, 0))  # Output: 3.0
```

---

## **2. Advanced Usage of `round()`**

### **Rounding Negative Numbers**

The behavior of `round()` is consistent for both positive and negative numbers.

```python
print(round(-3.14159, 2))  # Output: -3.14
print(round(-3.75))        # Output: -4
```

### **Rounding Ties (Halfway Cases)**

When a number is exactly halfway between two possible rounded values (e.g., `0.5`), Python uses **round half to even** (also known as **banker's rounding**) to avoid bias.

```python
# Example 1: Rounding 0.5 to the nearest integer
print(round(0.5))  # Output: 0 (rounds to even)

# Example 2: Rounding 1.5 to the nearest integer
print(round(1.5))  # Output: 2 (rounds to even)

# Example 3: Rounding 2.5 to the nearest integer
print(round(2.5))  # Output: 2 (rounds to even)
```

### **Rounding with `Decimal` for Precision**

For precise rounding, especially in financial calculations, you can use the `Decimal` module.

```python
from decimal import Decimal, ROUND_HALF_UP

# Example: Rounding with Decimal
number = Decimal('3.14159')
rounded = number.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)
print(rounded)  # Output: 3.14
```

---

## **3. Practical Use Cases**

### **Financial Calculations**

Rounding is often used in financial calculations to ensure results are in the correct currency format.

```python
# Example: Rounding to 2 decimal places for currency
price = 19.98765
rounded_price = round(price, 2)
print(f"Price: ${rounded_price}")  # Output: Price: $19.99
```

### **Formatting Output**

Rounding is useful for formatting output to a specific number of decimal places.

```python
# Example: Formatting a floating-point number
pi = 3.141592653589793
print(f"Pi rounded to 3 decimal places: {round(pi, 3)}")  # Output: Pi rounded to 3 decimal places: 3.142
```

### **Reducing Floating-Point Precision**

Rounding can help reduce the precision of floating-point numbers to avoid unnecessary decimal places.

```python
# Example: Reducing precision
value = 0.123456789
rounded_value = round(value, 4)
print(rounded_value)  # Output: 0.1235
```

---

## **4. Advanced Topics**

### **Custom Rounding Functions**

You can create custom rounding functions to implement specific rounding rules.

```python
# Example: Custom rounding function (always round up)
import math

def round_up(number, decimals=0):
    multiplier = 10 ** decimals
    return math.ceil(number * multiplier) / multiplier

print(round_up(3.14159, 2))  # Output: 3.15
```

### **Rounding with `math` Module**

The `math` module provides additional rounding functions like `math.floor()` and `math.ceil()`.

```python
import math

# Example 1: Round down using math.floor()
print(math.floor(3.75))  # Output: 3

# Example 2: Round up using math.ceil()
print(math.ceil(3.25))   # Output: 4
```

### **Performance Considerations**

- `round()` is fast and efficient for most use cases.
- For high-precision rounding, prefer the `Decimal` module.

---

## **Summary**

| Feature                    | Example                                                 |
| -------------------------- | ------------------------------------------------------- |
| Round to nearest integer   | `round(3.75)` → `4`                                     |
| Round to 2 decimal places  | `round(3.14159, 2)` → `3.14`                            |
| Round negative numbers     | `round(-3.75)` → `-4`                                   |
| Round ties (halfway cases) | `round(0.5)` → `0` (banker's rounding)                  |
| Round with `Decimal`       | `Decimal('3.14159').quantize(Decimal('0.01'))` → `3.14` |
| Custom rounding function   | `round_up(3.14159, 2)` → `3.15`                         |

---

This covers everything from basic to advanced usage of the `round()` function in Python. Let me know if you have further questions!
